jwst/firstframe/firstframe_step.py

```diff
@@ -19,7 +18,7 @@ class FirstFrameStep(Step):
     def process(self, input):
 
         # Open the input data model
-        with datamodels.open(input) as input_model:
+        with use_datamodel(input) as input_model:
```

Comment:
I'm going to use this step as an example but the patterns is repeated for several other steps in detector1.

In addition to the extra "clone" here, this step also copies the input model within the `do_correction` call:
https://github.com/spacetelescope/jwst/blob/c79c4aedb255de3f99357ab35c650459454d8b39/jwst/firstframe/firstframe_sub.py#L38

Since this PR is allowing steps to modify input (which I am on board with) removing this copy should further reduce memory usage. Same for line 31 below.

Reply:
I think I got all the unnecessary copies. Can you re-do your memory pressure test?

Reply:
With the current changes the memory consumption is max 17.9 GB for the small test file you shared with me. For comparison main was 20.3 GB and a previous version of this PR was 17.7 GB. See https://github.com/spacetelescope/jwst/pull/8588#discussion_r1653400731 for some details on why this most recent run is taking ~200 MB more.

Have you been able to run memray for your local testing? I'd be happy to help if you're running into issues.

Reply:
Here are a few more copies to investigate (note that some of these are in stcal):
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/rscd/rscd_sub.py#L91
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/saturation/saturation.py#L54
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/dq_init/dq_initialization.py#L62
- https://github.com/spacetelescope/stcal/blob/3fa4d42183fb384166d51f1aca8dd7bbe0ca6603/src/stcal/dark_current/dark_sub.py#L358
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/linearity/linearity.py#L16
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/reset/reset_sub.py#L59
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/ipc/ipc_corr.py#L78
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/lastframe/lastframe_sub.py#L36
- https://github.com/spacetelescope/jwst/blob/0b16f5451d502c666af0c1efc4e90070900f0789/jwst/firstframe/firstframe_sub.py#L38

These all pop up during the run with the test file, each contributing ~1.1GB to the memory usage.

Reply:
after our conversation this morning, I was told that we don't want the input model to be modified in-place, so there are some copies that need to remain for now.